worker/caasapplicationprovisioner/application.go

```diff
@@ -103,82 +108,134 @@ func (a *appWorker) loop() error {
 		return errors.Errorf("charm version 2 or greater required")
 	}
 
+	// Update the password once per worker start to avoid it changing to frequently.
+	a.password, err = utils.RandomPassword()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = a.facade.SetPassword(a.name, a.password)
+	if err != nil {
+		return errors.Annotate(err, "failed to set application api passwords")
+	}
+
 	// TODO(embedded): support more than statefulset
 	app := a.broker.Application(a.name, caas.DeploymentStateful)
 
 	var appLife life.Value
 	var appChanges watcher.NotifyChannel
 	var replicaChanges watcher.NotifyChannel
+	var appStateChanges watcher.NotifyChannel
 	var lastReportedStatus map[string]status.StatusInfo
 
-	for {
-		select {
-		case <-a.catacomb.Dead():
-			return a.catacomb.Err()
-		case <-a.changes:
-			appLife, err = a.facade.Life(a.name)
-			if errors.IsNotFound(err) {
-				appLife = life.Dead
-			} else if err != nil {
+	done := false
+
+	handleChange := func() error {
+		appLife, err = a.facade.Life(a.name)
+		if errors.IsNotFound(err) {
+			appLife = life.Dead
+		} else if err != nil {
+			return errors.Trace(err)
+		}
+		switch appLife {
+		case life.Alive:
+			if appStateChanges == nil {
+				appStateWatcher, err := a.facade.WatchApplication(a.name)
+				if err != nil {
+					return errors.Annotatef(err, "failed to watch for changes to application %q", a.name)
+				}
+				err = a.catacomb.Add(appStateWatcher)
+				if err != nil {
+					appStateWatcher.Kill()
```

Comment:
This isn't needed. The only error is ErrDying and the catacomb stops the watcher in that case